### DISCUSSION TITLE
Row deletion event

### DISCUSSION THREAD

**Discussion by ImogenBits** [sqlalchemy/sqlalchemy#9428] [Usage Questions] [answered]
I have a table that is related to a bunch of other tables and each row is owned by one entry in one of those other tables. A simple example is something like this:
```python
class Child(Base):
  id: Mapped[UUID]

class Parent1(Base):
  id: Mapped[UUID]
  child_id: Mapped[UUID] = mapped_column(ForeignKey("files.id"), init=False)
  child: Mapped[Child] = relationship(cascade="all, delete-orphan", single_parent=True, lazy="selectin")

class Parent2(Base):
  id: Mapped[UUID]
  child1_id: Mapped[UUID] = mapped_column(ForeignKey("files.id"), init=False)
  child1: Mapped[Child] = relationship(foreign_keys=child1_id, cascade="all, delete-orphan", single_parent=True, lazy="selectin")
  child2_id: Mapped[UUID] = mapped_column(ForeignKey("files.id"), init=False)
  child2: Mapped[Child] = relationship(foreign_keys=child2_id, cascade="all, delete-orphan", single_parent=True, lazy="selectin")
```

I now need to run additional code whenever a column in `Child` is removed (by setting the relationship of the parent to some other value) or a commit that inserted it is rolled back. I tried to get that to work by listening to the `deleted_to_detached` and `after_rollback` events but it's not working reliably. Are there other events I should be using instead?

I've set the lazy loading strategy to be eager to guarantee an orm object always is generated but that still doesn't seem to catch every instance of a row being deleted. And this is less critical, but since `after_rollback` doesn't tell you what actually was rolled back I need to track that manually, which seems like I'm missing the actually intended functionality somewhere?

**Comment by zzzeek** [comment, marked as answer]
"delete" in SQLAlchemy terminology means a DELETE statement is emitted to the database.  For this you'd normally use [after_delete](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.after_delete).

however your mention of "lazy loading" and "rollback", which don't "delete" anything, makes me think you actually mean "expire" and not "delete".   We don't have an "expired" event, but instead if what you are really doing is some kind of "load" of an attribute in response to other attributes being loaded, we have a suite of events intended to cover every one of these cases, which include [refresh](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.refresh), [refresh_flush](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.refresh_flush), [load](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.load), [unpickle](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.unpickle).

**Reply by ImogenBits** [reply]
yeah I'm really unsure about the correct terminology, sorry. Basically what I want is for the database rows to track files in the filesystem. I thought the correct way to go about that would be to write them at the `pending_to_persistent` event and then remove them if the session that wrote them gets rolled back, or if a session that emitted a DELETE statement gets committed. I thought that the `after_delete` event would fire before it actually gets committed, is that not the case?

**Reply by zzzeek** [reply]
if....you want the files to have data that's in the database, you'd normally write that after the database rows are committed.  that way you would not have to worry about rollback.   you would usually use after_commit() for that.    gather up the changes that will be written to the files either using the before_flush() event, or the per-object hooks like after_insert(), after_update(), after_delete().  by "gather up" I mean write some kind of data, like diffs or individual lines to be added/removed/whatever, into a collection (or perhaps a queue of sorts), which you can put in a place like [Session.info](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.info).  Then in after_commit() consume the contents of this queue and write out the changes to the files.

**Reply by ImogenBits** [reply]
right, that makes sense, thanks so much for the help!

One last question though, the after insert/delete events give me a Connection object, but the after commit one a Session, which means I can't get it to work using the info dicts. Is there a trick with getting the Session from the connection?

**Reply by zzzeek** [reply]
get it from the instance that's passed, `inspect(instance).session`

those particular event hooks pre-date the session itself

**Reply by ImogenBits** [reply]
awesome, thanks so much for the help!
